### PR TITLE
[security] Adding docstrings and type hints

### PR DESCRIPTION
**Pull request by john-bodley**
### CATEGORY

Choose one

- [ ] Bug Fix
- [ ] Enhancement (new features, refinement)
- [x] Refactor
- [ ] Add tests
- [ ] Build / Development Environment
- [ ] Documentation

### SUMMARY

This PR adds doc-strings and type hints to the security manager to help to provide more context and consistency. Note I've tried to ensure that the code remains unchanged however: 

1. I renamed a couple of methods/variables for accuracy. Note further polish could be applied. 
2. Given the vast number of methods in the `SupersetSecurityManager` (it's becoming almost intractable) I renamed private methods with a leading `_` (underscore). 
3. The Mypy type hints surfaced a couple of small issues during linting. 

A few notes/observations: 

1. I get the sense there's too many surfaces where security datasource checks happen (many occur on the view) and thus we susceptible for potentially forgetting to add them for new routes. Moving forward it seems like these checks should probably occur closer to the engine where appropriate.
2. I've tried to be as descriptive as possible but it took me quite some time to decipher what the various objects were. There may be merit in the future of being more explicit between Druid datasources and SQL tables. 
3. I think some of the private methods could probably be refactored. 
4. The SQL table names seem to be a mess, currently these can be defined as:

- `str`: `[[cluster.]schema.]table`
- `Tuple[str, str]`: `(schema, table)`
- `DatasourceName`

I believe there would be merit in aligning on one mechanism for defining SQL tables throughout Superset to help mitigate bugs/inaccuracies (note we've addresses issues related to this in the past). This is especially vital as it relates to security and a hardened security model should be one of Superset's core tenets.  A `dataclass` class may be a viable route. 

5. I'm not certain whether the `get_schema_perm` logic is correct as I believe that the `database` parameter can be `Union[Database, str]`. I'm not sure if that was intentional 

### TEST PLAN

CI.

### ADDITIONAL INFORMATION

- [ ] Has associated issue:
- [ ] Changes UI
- [ ] Requires DB Migration.
- [ ] Confirm DB Migration upgrade and downgrade tested.
- [ ] Introduces new feature or API
- [ ] Removes existing feature or API

### REVIEWERS

to: @DiggidyDave @etr2460 @michellethomas @mistercrunch @villebro

### CODE REVIEW

File: superset/security.py [hunk 1]
```diff
@@ -124,121 +131,276 @@ class SupersetSecurityManager(SecurityManager):
...
+            return tuple(table_name_pieces[1:])  # noqa: T484
         elif len(table_name_pieces) == 2:
-            return tuple(table_name_pieces)
+            return tuple(table_name_pieces)  # noqa: T484
         return (schema, table_name_pieces[0])
 
-    def datasource_access_by_fullname(self, database, table_in_query, schema):
-        table_schema, table_name = self.get_schema_and_table(table_in_query, schema)
-        return self.datasource_access_by_name(database, table_name, schema=table_schema)
+    def _datasource_access_by_fullname(
```
Comment (john-bodley) [inline superset/security.py]: Note I find the naming of this method somewhat confusing and there's also a method called `_datasource_access_by_name`, i.e, without the `full`.

Comment (DiggidyDave) [reply superset/security.py]: yeah this is super confusing. I'm probably an outlier among those reading this, but I don't know what "fallback sql schema" means, would it be appropriate to clarify that here in the param docs?

**Comment by DiggidyDave** [comment]
> I believe there would be merit in aligning on one mechanism for defining SQL tables throughout Superset ... A dataclass class may be a viable route.

👍 agreed! if something is important enough that folks have invented 3 ways of doing it, it probably should have a first-class type :-) 

Thanks for doing all of this work

File: superset/security.py (hunk 1 above)
Comment (john-bodley) [reply superset/security.py]: @DiggidyDave I added more documentation regarding the "fallback SQL schema".

File: superset/security.py [hunk 2]
```diff
@@ -30,11 +31,17 @@
...
 
 from superset import sql_parse
 from superset.connectors.connector_registry import ConnectorRegistry
 from superset.exceptions import SupersetSecurityException
-from superset.utils.core import DatasourceName
+
+if TYPE_CHECKING:
+    from superset.models.core import Database, BaseDatasource
+
+from superset.utils.core import DatasourceName  # noqa: I202
```
Comment (villebro) [inline superset/security.py]: Perhaps we should add ignoring of `I202` to `tox.ini`? @sturmer ran intro a few of those in #7643 and it seems `flake8` and `black` aren't aligned in their thinking on newlines.

File: superset/security.py [hunk 3]
```diff
@@ -124,121 +131,278 @@ class SupersetSecurityManager(SecurityManager):
...
         return self.schema_access(datasource) or self.can_access(
             "datasource_access", datasource.perm
         )
 
-    def get_datasource_access_error_msg(self, datasource):
-        return """This endpoint requires the datasource {}, database or
-            `all_datasource_access` permission""".format(
-            datasource.name
-        )
+    def get_datasource_access_error_msg(self, datasource: "BaseDatasource") -> str:
```
Comment (villebro) [inline superset/security.py]: Why is this in quotes?

Edit: googled it, learned something new 👍

**Review by villebro — APPROVED** [review]
Awesome work 👍 Apart from the small comment about `I202` LGTM.

File: superset/security.py (hunk 2 above)
Comment (john-bodley) [reply superset/security.py]: @villebro I can create another PR which ignores `I202` and removes all references.

**Pull request merged**